File: starters/L3.4/judge.py
"""
L3.4 starter -- LLM-as-judge eval runner.

Reads evals.csv with columns: category, input, expected_or_rubric, grader.
- grader=substring: simple contains-check (cheap, deterministic)
- grader=judge: send to a judge model with a rubric, expect 0/1/2 score

Run:
  python judge.py --threshold 0.85
Exit code 1 if below threshold -- wire this into CI.
"""
import argparse
import csv
import json
import sys
import anthropic

client = anthropic.Anthropic()

CANDIDATE_MODEL = "claude-haiku-4-5"
JUDGE_MODEL = "claude-sonnet-4-6"  # NOT Haiku -- too lenient as a judge

YOUR_PROMPT = (
    "Extract decisions, owners, and open questions from the text below. "
    "Return strict JSON only.\n\n"
    "TEXT:\n{input}"
)

JUDGE_PROMPT = """You are grading a model's output. Be strict.

Question to the model:
{input}

Rubric for a good answer:
{rubric}

Model output:
{output}

Reply with strict JSON only:
{{"score": <0|1|2>, "reason": "<one sentence>"}}

0 = wrong or misses the point. 1 = partially right. 2 = correct and well-formed.
"""


def run_candidate(input_text: str) -> str:
    resp = client.messages.create(
        model=CANDIDATE_MODEL,
        max_tokens=512,
        messages=[{"role": "user", "content": YOUR_PROMPT.format(input=input_text)}],
    )
    return resp.content[0].text


def grade_substring(output: str, expected: str) -> int:
    return 2 if expected.lower() in output.lower() else 0


def grade_judge(input_text: str, rubric: str, output: str) -> tuple[int, str]:
    resp = client.messages.create(
        model=JUDGE_MODEL,
        max_tokens=200,
        messages=[{"role": "user", "content": JUDGE_PROMPT.format(input=input_text, rubric=rubric, output=output)}],
    )
    text = resp.content[0].text.strip()
    try:
        verdict = json.loads(text)
        return int(verdict.get("score", 0)), verdict.get("reason", "")
    except Exception:
        return 0, f"judge returned unparseable response: {text[:120]}"


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--threshold", type=float, default=0.85)
    ap.add_argument("--evals", default="evals.csv")
    args = ap.parse_args()

    with open(args.evals, newline="") as f:
        rows = list(csv.DictReader(f))

    total = max_possible = 0
    with open("results.csv", "w", newline="") as out:
        w = csv.writer(out)
        w.writerow(["category", "grader", "score", "max", "input", "output", "reason"])
        for row in rows:
            output = run_candidate(row["input"])
            grader = row.get("grader", "substring")
            if grader == "judge":
                score, reason = grade_judge(row["input"], row["expected_or_rubric"], output)
                max_score = 2
            else:
                score = grade_substring(output, row["expected_or_rubric"])
                reason = ""
                max_score = 2
            total += score
            max_possible += max_score
            w.writerow([row["category"], grader, score, max_score, row["input"], output, reason])
            print(f"[{grader:9}] {score}/{max_score}  {row['category']:14} {row['input'][:60]}")

    pass_rate = total / max_possible
    print(f"\n{total}/{max_possible} points ({100*pass_rate:.1f}%) -- threshold {100*args.threshold:.0f}%")
    if pass_rate < args.threshold:
        print("FAIL")
        sys.exit(1)
    print("PASS")


if __name__ == "__main__":
    main()

File: starters/L3.4/judge.ts
/**
 * L3.4 starter -- LLM-as-judge eval runner.
 *
 * Reads evals.csv with columns: category, input, expected_or_rubric, grader.
 * - grader=substring: simple contains-check (cheap, deterministic)
 * - grader=judge: send to a judge model with a rubric, expect 0/1/2 score
 *
 * Run:
 *   npm install
 *   npx ts-node judge.ts --threshold 0.85
 * Exit code 1 if below threshold -- wire this into CI.
 */
import Anthropic from "@anthropic-ai/sdk";
import * as fs from "fs";

const client = new Anthropic();

const CANDIDATE_MODEL = "claude-haiku-4-5";
const JUDGE_MODEL = "claude-sonnet-4-6"; // NOT Haiku -- too lenient as a judge

const YOUR_PROMPT =
  "Extract decisions, owners, and open questions from the text below. " +
  "Return strict JSON only.\n\n" +
  "TEXT:\n{input}";

const JUDGE_PROMPT = `You are grading a model's output. Be strict.

Question to the model:
{input}

Rubric for a good answer:
{rubric}

Model output:
{output}

Reply with strict JSON only:
{"score": <0|1|2>, "reason": "<one sentence>"}

0 = wrong or misses the point. 1 = partially right. 2 = correct and well-formed.`;

interface EvalRow {
  category: string;
  input: string;
  expected_or_rubric: string;
  grader: string;
}

async function runCandidate(inputText: string): Promise<string> {
  const resp = await client.messages.create({
    model: CANDIDATE_MODEL,
    max_tokens: 512,
    messages: [{ role: "user", content: YOUR_PROMPT.replace("{input}", inputText) }],
  });
  return resp.content[0].type === "text" ? resp.content[0].text : "";
}

function gradeSubstring(output: string, expected: string): number {
  return output.toLowerCase().includes(expected.toLowerCase()) ? 2 : 0;
}

async function gradeJudge(
  inputText: string,
  rubric: string,
  output: string
): Promise<{ score: number; reason: string }> {
  const resp = await client.messages.create({
    model: JUDGE_MODEL,
    max_tokens: 200,
    messages: [
      {
        role: "user",
        content: JUDGE_PROMPT.replace("{input}", inputText)
          .replace("{rubric}", rubric)
          .replace("{output}", output),
      },
    ],
  });
  const text = (resp.content[0].type === "text" ? resp.content[0].text : "").trim();
  try {
    const verdict = JSON.parse(text) as { score: number; reason: string };
    return { score: verdict.score ?? 0, reason: verdict.reason ?? "" };
  } catch {
    return { score: 0, reason: `judge returned unparseable response: ${text.slice(0, 120)}` };
  }
}

function parseCsv(content: string): EvalRow[] {
  const lines = content.trim().split("\n");
  const headers = lines[0].split(",");
  return lines.slice(1).map((line) => {
    const vals = line.split(",");
    return {
      category: vals[headers.indexOf("category")] ?? "",
      input: vals[headers.indexOf("input")] ?? "",
      expected_or_rubric: vals[headers.indexOf("expected_or_rubric")] ?? "",
      grader: vals[headers.indexOf("grader")] ?? "substring",
    };
  });
}

(async () => {
  const args = process.argv.slice(2);
  const thresholdIdx = args.indexOf("--threshold");
  const threshold = thresholdIdx >= 0 ? parseFloat(args[thresholdIdx + 1]) : 0.85;
  const evalsIdx = args.indexOf("--evals");
  const evalsFile = evalsIdx >= 0 ? args[evalsIdx + 1] : "evals.csv";

  const rows = parseCsv(fs.readFileSync(evalsFile, "utf-8"));
  let total = 0;
  let maxPossible = 0;
  const csvRows = ["category,grader,score,max,input,output,reason"];

  for (const row of rows) {
    const output = await runCandidate(row.input);
    let score: number;
    let reason = "";
    const maxScore = 2;

    if (row.grader === "judge") {
      const verdict = await gradeJudge(row.input, row.expected_or_rubric, output);
      score = verdict.score;
      reason = verdict.reason;
    } else {
      score = gradeSubstring(output, row.expected_or_rubric);
    }

    total += score;
    maxPossible += maxScore;

    csvRows.push(
      [row.category, row.grader, score, maxScore, row.input, output, reason]
        .map((v) => `"${String(v).replace(/"/g, '""')}"`)
        .join(",")
    );
    console.log(
      `[${row.grader.padEnd(9)}] ${score}/${maxScore}  ${row.category.padEnd(14)} ${row.input.slice(0, 60)}`
    );
  }

  fs.writeFileSync("results.csv", csvRows.join("\n"));

  const passRate = total / maxPossible;
  console.log(
    `\n${total}/${maxPossible} points (${(100 * passRate).toFixed(1)}%) -- threshold ${(100 * threshold).toFixed(0)}%`
  );
  if (passRate < threshold) {
    console.log("FAIL");
    process.exit(1);
  }
  console.log("PASS");
})();
